FutureWarning: split() requires a non-empty pattern match

I found this while investigating astropy/astropy#7928 by:

* Un-silencing warnings in `pytest`.
* Turn warnings into exception (except for `ImportWarning` from Cython, which I ignored still).
* Run tests on `coordinates` (saw this in `convolution` too).

```
__________________ [doctest] docs/coordinates/solarsystem.rst __________________
.../pytest_doctestplus/plugin.py:122: in runtest
    raise_on_error=True, verbose=False, encoding='utf-8')
.../doctest.py:2073: in testfile
    test = parser.get_doctest(text, globs, name, filename, 0)
.../doctest.py:669: in get_doctest
    return DocTest(self.get_examples(string, name), globs,
.../doctest.py:683: in get_examples
    return [x for x in self.parse(string, name)
.../pytest_doctestplus/plugin.py:192: in parse
    required = re.split(r'\s*,?\s*', match.group(1))
.../re.py:212: in split
    return _compile(pattern, flags).split(string, maxsplit)
E   FutureWarning: split() requires a non-empty pattern match.
```

https://github.com/astropy/pytest-doctestplus/blob/6703a182cf931968b5a9f2ac1949059270a88002/pytest_doctestplus/plugin.py#L213-L214

Here is my `pytest` header:
```
platform linux -- Python 3.6.4, pytest-3.2.1, py-1.4.34, pluggy-0.4.0

Running tests with Astropy version 3.1.dev22915.
Running tests in astropy/coordinates docs/coordinates.

Date: 2018-10-26T15:57:40

Platform: Linux-3.10.0-862.14.4.el7.x86_64-x86_64-with-redhat-7.5-Maipo

Executable: .../envs/py36/bin/python

Full Python Version: 
3.6.4 |Anaconda, Inc.| (default, Jan 16 2018, 18:10:19) 
[GCC 7.2.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Numpy: 1.15.2
Scipy: 1.1.0
Matplotlib: 2.2.2
h5py: 2.8.0
Pandas: 0.20.3
Asdf: 2.0.1
Cython: 0.26.1
astropy_helpers: 2.0.6
Using Astropy options: remote_data: any.

PyQt5 5.6 -- Qt runtime 5.6.2 -- Qt compiled 5.6.2
rootdir: /tmp/astropy-test-woysnyxx/lib/python3.6/site-packages, inifile: setup.cfg
plugins: xdist-1.17.1, qt-2.4.1, openfiles-0.2.0, mpl-0.8, flake8-0.8.1, doctestplus-0.1.2, arraydiff-0.1, remotedata-0.2.2.dev0, ci-watson-0.0.0.dev33+gdec4b009
```

--- a/pytest_doctestplus/plugin.py
+++ b/pytest_doctestplus/plugin.py
@@ -12,10 +12,9 @@ import re
 import sys
 import warnings
 
-from distutils.version import LooseVersion
-
 import pytest
 
+from pytest_doctestplus.utils import ModuleChecker
 from .output_checker import OutputChecker, FIX
 
 comment_characters = {'txt': '#',
@@ -395,32 +394,26 @@ class DocTestFinderPlus(doctest.DocTestFinder):
 
     # Caches the results of import attempts
     _import_cache = {}
+    _module_checker = ModuleChecker()
 
     @classmethod
     def check_required_modules(cls, mods):
+        """
+        Check that modules in :attr:`mods` list are available.
+        :param mods: list of module names. Modules can have specified versions
+          (eg 'numpy>=1.15')
+        :return: True if all modules are available.
+        """
         for mod in mods:
             if mod in cls._import_cache:
                 if not cls._import_cache[mod]:
                     return False
 
-            if LooseVersion(sys.version) < LooseVersion('3.4'):
-                import imp
-                try:
-                    module = imp.find_module(mod)
-                except ImportError:
-                    module = None
+            if cls._module_checker.check(mod):
+                cls._import_cache[mod] = True
             else:
-                import importlib.util
-                try:
-                    module = importlib.util.find_spec(mod)
-                except ImportError:
-                    module = None
-
-            if module is None:
                 cls._import_cache[mod] = False
                 return False
-            else:
-                cls._import_cache[mod] = True
         return True
 
     def find(self, obj, name=None, module=None, globs=None,

--- a/pytest_doctestplus/plugin.py
+++ b/pytest_doctestplus/plugin.py
@@ -256,7 +256,8 @@ def pytest_configure(config):
                         match = matches[0]
 
                     if match:
-                        required = re.split(r'\s*,?\s*', match.group(1))
+                        # 'a a' or 'a,a' or 'a, a'-> [a, a]
+                        required = re.split(r'\s*[,\s]\s*', match.group(1))
                 elif isinstance(entry, doctest.Example):
                     if (skip_all or skip_next or
                         not DocTestFinderPlus.check_required_modules(required)):

--- a/pytest_doctestplus/plugin.py
+++ b/pytest_doctestplus/plugin.py
@@ -398,11 +398,17 @@ class DocTestFinderPlus(doctest.DocTestFinder):
 
     @classmethod
     def check_required_modules(cls, mods):
-        """
-        Check that modules in :attr:`mods` list are available.
-        :param mods: list of module names. Modules can have specified versions
-          (eg 'numpy>=1.15')
-        :return: True if all modules are available.
+        """Check that modules in `mods` list are available.
+
+        Parameters
+        ----------
+        mods : list of str
+            List of modules. Modules can have specified versions (eg 'numpy>=1.15')
+
+        Returns
+        -------
+        bool
+            True if all modules in list are available.
         """
         for mod in mods:
             if mod in cls._import_cache:

--- a/pytest_doctestplus/utils.py
+++ b/pytest_doctestplus/utils.py
@@ -49,7 +49,7 @@ class ModuleChecker:
         match = re.match(r'([A-Za-z0-9-_]+)([^A-Za-z0-9-_]+)([\d.]+$)', module)
         if not match:
             return False
-        package, cmp, version = match[1], match[2], match[3]
+        package, cmp, version = match.groups()
         package = package.lower()
 
         if package in self.packages:

--- a/pytest_doctestplus/utils.py
+++ b/pytest_doctestplus/utils.py
@@ -1,0 +1,93 @@
+import logging
+import operator
+import re
+import subprocess
+import sys
+from distutils.version import LooseVersion
+
+logger = logging.getLogger(__name__)
+
+
+class ModuleChecker:
+    def __init__(self):
+        if LooseVersion(sys.version) < LooseVersion('3.4'):
+            import imp
+            self._find_module = imp.find_module
+            self._find_distribution = self._check_distribution
+            self.packages = self.get_packages()
+        else:
+            import importlib.util
+            import pkg_resources
+            self._find_module = importlib.util.find_spec
+            self._find_distribution = pkg_resources.require
+            self.packages = {}
+
+    def get_packages(self):
+        packages = subprocess.check_output([sys.executable, '-m', 'pip', 'freeze']).decode().splitlines()
+        packages = [package.split('==') for package in packages if '==' in package]
+        return {name.lower(): version for name, version in packages}
+
+    def compare_versions(self, v1, v2, op):
+        op_map = {
+            '<': operator.lt,
+            '<=': operator.le,
+            '>': operator.gt,
+            '>=': operator.ge,
+            '==': operator.eq,
+        }
+        if op not in op_map:
+            return False
+        op = op_map[op]
+        return op(LooseVersion(v1), LooseVersion(v2))
+
+    def _check_distribution(self, module):
+        """
+        Python 2 (and <3.4) compatible version of pkg_resources.require.
+        But unlike pkg_resources.require it just checks whether package is
+        installed and has required version.
+        """
+        match = re.match(r'([A-Za-z0-9-_]+)([^A-Za-z0-9-_]+)([\d.]+$)', module)
+        if not match:
+            return False
+        package, cmp, version = match[1], match[2], match[3]
+        package = package.lower()
+
+        if package in self.packages:
+            installed_version = self.packages[package]
+            if self.compare_versions(installed_version, version, cmp):
+                return True
+            else:
+                logger.warning(
+                    "{} {} is installed. Version {}{} is required".format(
+                        package, installed_version, cmp, version
+                    )
+                )
+                return False
+        logger.warning("The '{}' distribution was not found and is required by the application".format(package))
+        return False
+
+    def find_module(self, module):
+        """Search for modules specification."""
+        try:
+            return self._find_module(module)
+        except ImportError:
+            return None
+
+    def find_distribution(self, dist):
+        """Search for distribution with specified version (eg 'numpy>=1.15')."""
+        try:
+            return self._find_distribution(dist)
+        except Exception as e:
+            logger.warning(e)
+            return None
+
+    def check(self, module):
+        """
+        Return True if module with specified version exists.
+        >>> ModuleChecker().check('foo>=1.0')
+        False
+        >>> ModuleChecker().check('pytest>1.0')
+        True
+        """
+        mods = self.find_module(module) or self.find_distribution(module)
+        return bool(mods)

--- a/tests/docs/skip_some.rst
+++ b/tests/docs/skip_some.rst
@@ -64,3 +64,19 @@ run:
     >>> import foobar
     >>> foobar.baz(42)
     1
+
+
+Package version
+===============
+
+Code in doctest should run only if version condition is satisfied:
+
+.. doctest-requires:: numpy<=0.1
+
+    >>> import numpy
+    >>> assert 0
+
+
+.. doctest-requires:: pytest>=1.0 pytest>=2.0
+
+    >>> import pytest

--- a/tests/test_doctestplus.py
+++ b/tests/test_doctestplus.py
@@ -360,7 +360,7 @@ def test_requires(testdir):
     )
     testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(passed=1)
 
-    # testing this in case if doctest-requires just ignores everything
+    # testing this in case if doctest-requires just ignores everything and pass unconditionally
     p = testdir.makefile(
         '.rst',
         """
@@ -370,3 +370,25 @@ def test_requires(testdir):
         """
     )
     testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(failed=1)
+
+    # package with version is available
+    p = testdir.makefile(
+        '.rst',
+        """
+        .. doctest-requires:: sys pytest>=1.0
+            >>> import sys, pytest
+        """
+    )
+    testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(passed=1)
+
+    # package with version is not available
+    p = testdir.makefile(
+        '.rst',
+        """
+        .. doctest-requires:: sys pytest<1.0 glob
+            >>> import sys, pytest, glob
+            >>> assert 0
+        """
+    )
+    # passed because 'pytest<1.0' was not satisfied and 'assert 0' was not evaluated
+    testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(passed=1)

--- a/tests/test_doctestplus.py
+++ b/tests/test_doctestplus.py
@@ -331,3 +331,42 @@ class TestFloats:
             [str(i) for i in range(20)],
             [[], ['1', '2'], ['7'], ['6'], []],
         )
+
+
+def test_requires(testdir):
+    testdir.makeini(
+        """
+        [pytest]
+        doctestplus = enabled
+    """)
+
+    # should be ignored
+    p = testdir.makefile(
+        '.rst',
+        """
+        .. doctest-requires:: foobar
+            >>> import foobar
+        """
+    )
+    testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(passed=1)
+
+    # should run as expected
+    p = testdir.makefile(
+        '.rst',
+        """
+        .. doctest-requires:: sys
+            >>> import sys
+        """
+    )
+    testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(passed=1)
+
+    # testing this in case if doctest-requires just ignores everything
+    p = testdir.makefile(
+        '.rst',
+        """
+        .. doctest-requires:: sys glob, re,math
+            >>> import sys
+            >>> assert 0
+        """
+    )
+    testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(failed=1)

--- a/tests/test_utils.py
+++ b/tests/test_utils.py
@@ -1,0 +1,25 @@
+from pytest_doctestplus.utils import ModuleChecker
+
+
+class TestModuleChecker:
+    def test_simple(self):
+        c = ModuleChecker()
+        assert c.check('sys')
+        assert not c.check('foobar')
+
+    def test_with_version(self):
+        c = ModuleChecker()
+        assert c.check('pytest>1.0')
+        assert not c.check('foobar>1.0')
+
+    def test_check_distribution(self):
+        c = ModuleChecker()
+        # in python3.4+ packages attribute will not be populated
+        # because it calls 'pip freeze' which is slow
+        if not c.packages:
+            c.packages = c.get_packages()
+            # after this we will be able to test _check_distribution even in
+            # python3.4+ environment
+        assert c._check_distribution('pytest>1.0')
+        assert not c._check_distribution('pytest<1.0')
+        assert not c._check_distribution('foobar>1.0')

